front/lib/resources/membership_resource.ts

```diff
@@ -0,0 +1,344 @@
+        }
+      : {};
+
+    where.workspaceId = workspace.id;
+
+    return MembershipModel.count({
+      where,
+      distinct: true,
+      col: "userId",
+    });
+  }
+
+  static async createMembership({
+    userId,
+    workspace,
+    role,
+    startAt = new Date(),
+  }: {
+    userId: number;
+    workspace: WorkspaceType | Workspace;
+    role: MembershipRoleType;
+    startAt?: Date;
+  }): Promise<MembershipResource> {
+    if (startAt > new Date()) {
+      throw new Error("Cannot create a membership in the future");
+    }
+    if (
+      await MembershipModel.count({
+        where: {
+          userId,
+          workspaceId: workspace.id,
+          endAt: {
+            [Op.or]: [{ [Op.eq]: null }, { [Op.gt]: startAt }],
+          },
+        },
+      })
+    ) {
+      throw new Error(
+        `User ${userId} already has an active membership in workspace ${workspace.id}`
+      );
+    }
+    const newMembership = await MembershipModel.create({
+      startAt,
+      userId,
+      workspaceId: workspace.id,
+      role,
+    });
+
+    return new MembershipResource(MembershipModel, newMembership.get());
+  }
+
+  static async revokeMembership({
+    userId,
+    workspace,
+    endAt = new Date(),
+  }: {
+    userId: number;
+    workspace: WorkspaceType | Workspace;
+    endAt?: Date;
+  }): Promise<
+    Result<
+      undefined,
+      {
+        type: "not_found" | "already_revoked";
+      }
+    >
+  > {
+    const membership = await this.getLatestMembershipOfUserInWorkspace({
+      userId,
+      workspace,
+    });
+    if (!membership) {
+      return new Err({ type: "not_found" });
+    }
+    if (endAt < membership.startAt) {
+      throw new Error("endAt must be after startAt");
+    }
+    if (membership.endAt) {
+      return new Err({ type: "already_revoked" });
+    }
+    await MembershipModel.update({ endAt }, { where: { id: membership.id } });
+    return new Ok(undefined);
+  }
+
+  static async updateMembershipRole({
+    userId,
+    workspace,
+    newRole,
+    allowTerminated = false,
+  }: {
+    userId: number;
+    workspace: WorkspaceType | Workspace;
+    newRole: Exclude<MembershipRoleType, "revoked">;
+    // If true, allow updating the role of a terminated membership (which will also un-terminate it).
+    allowTerminated?: boolean;
+  }): Promise<
+    Result<
+      void,
+      {
+        type: "not_found" | "already_on_role" | "membership_already_terminated";
+      }
+    >
+  > {
+    const membership = await this.getLatestMembershipOfUserInWorkspace({
+      userId,
+      workspace,
+    });
+    if (membership?.endAt && !allowTerminated) {
+      return new Err({ type: "membership_already_terminated" });
+    }
+    if (!membership) {
+      return new Err({ type: "not_found" });
+    }
+
+    // If the membership is not terminated, we update the role in place.
+    // TODO(@fontanierh): check if we want to terminate + create a new membership with new role instead ?
```

Comment:
Debatable if we want to do this or not. I went with updating in place since we don't have a clear use-case for keep history of the roles (and it's simpler and more efficient) but could be convinced otherwise.